src/host/inputBuffer.cpp

```diff
@@ -407,24 +410,45 @@ void InputBuffer::_ReadBuffer(_Out_ std::deque<std::unique_ptr<IInputEvent>>& ou
             }
         }
 
-        if (performNormalRead)
-        {
-            readEvents.push_back(std::move(_storage.front()));
-            _storage.pop_front();
-        }
-
-        ++virtualReadCount;
+        // GH #8663: Before we read this key from the buffer, check that there's
+        // space for it. If we're calling Read without unicode being set, then I
+        // believe we're also going to try and break this key event into one key
+        // for each OEM character. Problem is though, one unicode codepoint can
+        // be more than two chars long. So don't just use IsGlyphFullWidth,
+        // actually check with the codepoint how wide this key should be.
+        size_t sizeOfThisKey = 1;
+        const auto& keyToRead{ _storage.front() };
         if (!unicode)
         {
-            if (readEvents.back()->EventType() == InputEventType::KeyEvent)
+            if (keyToRead->EventType() == InputEventType::KeyEvent)
             {
-                const KeyEvent* const pKeyEvent = static_cast<const KeyEvent* const>(readEvents.back().get());
-                if (IsGlyphFullWidth(pKeyEvent->GetCharData()))
+                const KeyEvent* const pKeyEvent = static_cast<const KeyEvent* const>(keyToRead.get());
+
+                // convert from wchar to char
+                std::wstring wstr{ pKeyEvent->GetCharData() };
+                const auto str = ConvertToA(codepage, wstr);
+                sizeOfThisKey = str.size();
+                if (virtualReadCount + sizeOfThisKey > readCount)
                 {
-                    ++virtualReadCount;
+                    break;
                 }

```

Comment:
This loop now has two bounds checks instead of one, which might be a bit confusing. We can simplify it by changing the loop to be `while (!_storage.empty())` and replacing this if condition with the following:
```cpp
if (!unicode)
{
    ...
}

virtualReadCount += sizeOfThisKey;
if (virtualReadCount > readCount)
{
    break;
}

if (performNormalRead)
{
    ...
}
```